File: ocaml/lib/bsearch_benchmark.ml
type 'a tree = Leaf | Node of 'a * 'a tree * 'a tree

let rec insert n = function
  | Leaf -> Node (n, Leaf, Leaf)
  | Node (v, l, r) ->
      if n < v then Node (v, insert n l, r)
      else if n > v then Node (v, l, insert n r)
      else Node (v, l, r)

let rec bsearch_twocomp n = function
  | Leaf -> false
  | Node (v, l, r) ->
      if n < v then bsearch_twocomp n l
      else if n > v then bsearch_twocomp n r
      else true

let bsearch_onecomp n t =
  let rec bsearch_helper target t cand =
    match (t, cand) with
    | Leaf, None -> false
    | Leaf, Some v -> v = target
    | Node (v, l, r), c ->
        if target < v then bsearch_helper target l c
        else bsearch_helper target r (Some v)
  in
  bsearch_helper n t None

(* Worst case is an unbalanced tree where n > v *)

let build_tree size =
  let rec helper t = function
    | 0 -> t
    | n -> helper (insert (Random.int size) t) (n - 1)
  in
  helper Leaf size

File: go/bsearch_benchmark.go
package bsearch_benchmark

import "math/rand"

type Tree struct {
	v int
	l *Tree
	r *Tree
}

func BSearchOneComp(t *Tree, n int) bool {
	var candidate *Tree
	for t != nil {
		if n < t.v {
			t = t.l
		} else {
			candidate = t
			t = t.r
		}
	}
	if candidate != nil && candidate.v == n {
		return true
	}

	return false
}

func BSearchTwoComp(t *Tree, n int) bool {
	for t != nil {
		if n < t.v {
			t = t.l
		} else if n > t.v {
			t = t.r
		} else {
			return true
		}

	}
	return false
}

func Insert(t *Tree, n int) {
	for t != nil {
		if n < t.v {
			if t.l == nil {
				t.l = &Tree{n, nil, nil}
				return
			}
			t = t.l
		} else if n > t.v {
			if t.r == nil {
				t.r = &Tree{n, nil, nil}
				return
			}
			t = t.r
		} else {
			return
		}
	}

	panic("can't insert into nil tree")

}

func BuildTree(n int) *Tree {
	t := &Tree{rand.Intn(n), nil, nil}
	for i := 0; i < n; i++ {
		Insert(t, rand.Intn(n))

	}
	return t
}
